reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -16,349 +16,205 @@
-            assert(NULL != sa_hmac_key);
-            sa_status status_hmac = sa_key_import(sa_hmac_key, SA_KEY_FORMAT_EXPORTED,
-                                  nflxProvision->hmacKey,nflxProvision->hmacKeyLength,
-                                  NULL);
 
-            sa_key *sa_wrapping_key = create_uninitialized_sa_key();
-            assert(NULL != sa_wrapping_key);
-
-            sa_status status_wrapping = sa_key_import(sa_wrapping_key, SA_KEY_FORMAT_EXPORTED,
-                                  nflxProvision->wrappingKey,nflxProvision->wrappingKeyLength,
-                                  NULL);
-
-            if(NULL != nflxProvision->hmacKey) {
-               free(nflxProvision->hmacKey);
-               nflxProvision->hmacKey = NULL;
-            }
-            if(NULL != nflxProvision->wrappingKey) {
-               free(nflxProvision->wrappingKey);
-               nflxProvision->wrappingKey = NULL;
-            }
-            if(NULL != nflxProvision->esnContainer) {
-               free(nflxProvision->esnContainer);
-               nflxProvision->esnContainer = NULL;
-            }
-            if(NULL != nflxProvision) {
-               free(nflxProvision);
-               nflxProvision = NULL;
-            }
-            release_sa_key(sa_hmac_key);
-            release_sa_key(sa_wrapping_key);
-
-            if(SA_STATUS_OK != status_hmac) {
-               ERROR("Faild to import netflix hmac key");
-               return status_hmac;
-            }
-            if(SA_STATUS_OK != status_wrapping) {
-               ERROR("Faild to import netflix wrapping key");
-            }
-            return status_wrapping;
+      case NETFLIX_PROVISIONING:
+         if (NULL == ((NetflixProvisioning*)in)->hmacKey         ||
+             0 == ((NetflixProvisioning*)in)->hmacKeyLength      ||
+             NULL == ((NetflixProvisioning*)in)->wrappingKey     ||
+             0 == ((NetflixProvisioning*)in)->wrappingKeyLength  ||
+             NULL == ((NetflixProvisioning*)in)->esnContainer    ||
+             0 == ((NetflixProvisioning*)in)->esnContainerLength) {
+             ERROR("null input or input size is  0");
+             return SA_STATUS_NULL_PARAMETER;
          }
          break;
-      default:
-         break;
-  }
-
 
-  return status;
+      case APPLE_MFI_PROVISIONING:
+         if (NULL == ((AppleMfiProvisioning*)in)->mfiBaseKey            ||
+             0 == ((AppleMfiProvisioning*)in)->mfiBaseKeyLength         ||
+             NULL == ((AppleMfiProvisioning*)in)->mfiProvisioningObject ||
+             0 == ((AppleMfiProvisioning*)in)->mfiProvisioningObjectLength) {
+             ERROR("null private key/object or size is  0");
+             return SA_STATUS_NULL_PARAMETER;
+         }
+         /*TODO SoC Vendor:here just provide an example to pass-through MFI base key
+           and provisioning obejet,
+           how to use these data is all up to SOC vendor.
+          */
+        {
```

Comment:
"TODO SoC Vendor" code all needs to be in an "src/internal" or "src/porting" file.   Code that is not located in internal or porting should not be changed and is considered platform independent.

This comment might be unnecessary, as I don't know that the SOC vendor would ever change the code here, as the data structures are all defined in the spec.